File: 0515-find-largest-value-in-each-tree-row/0515-find-largest-value-in-each-tree-row.cpp

class Solution {
public:
//     vector<int> largestValues(TreeNode* root) {
//         queue<TreeNode*> q;
//         vector<int> values;
        
//         if(root != NULL) q.push(root);
        
//         while(!q.empty()) {
//             int maxi =INT_MIN, n = q.size();
//             for(int i = 0; i < n; i++) {
//                 TreeNode* node = q.front();
//                  q.pop();
//                 maxi = max(maxi, node->val);
//                 if(node->left != NULL) q.push(node->left);
//                 if(node->right != NULL) q.push(node->right);
//             }
//             values.push_back(maxi);
//         }
        
//         return values;
//     }
    public:
    vector<int> largestValues(TreeNode* root) {
        vector<int> res;
        dfs(root, 1, res);
        return res;
    }

private:
    void dfs(TreeNode* node, int depth, vector<int>& res) {
        if (!node) {
            return;
        }
        dfs(node->left, depth + 1, res);
        dfs(node->right, depth + 1, res);
        if (depth > res.size()) {
            res.resize(depth, INT_MIN);
        }
        res[depth - 1] = max(res[depth - 1], node->val);
    }
};

File: 0515-find-largest-value-in-each-tree-row/0515-find-largest-value-in-each-tree-row.java

class Solution {
    public List<Integer> largestValues(TreeNode root) {
        Queue<TreeNode> q = new LinkedList<TreeNode>();
        List<Integer> values = new ArrayList<Integer>();
        
        if(root != null) q.offer(root);
        
        while(!q.isEmpty()) {
            int max = Integer.MIN_VALUE, n = q.size();
            for(int i = 0; i < n; i++) {
                TreeNode node = q.poll();
                max = Math.max(max, node.val);
                if(node.left != null) q.offer(node.left);
                if(node.right != null) q.offer(node.right);
            }
            values.add(max);
        }
        
        return values;
    }
}